test/tools/cmap_spec_runner.ts

```diff
@@ -0,0 +1,336 @@
+  | { name: 'checkOut'; thread: string; label: string }
+  | { name: 'checkIn'; connection: string }
+  | { name: 'clear' | 'close' | 'ready' };
+
+type cmapPoolOptions = {
+  maxPoolSize?: number;
+  minPoolSize?: number;
+  maxIdleTimeMS?: number;
+  waitQueueTimeoutMS?: number;
+};
+
+type cmapEvent = {
+  type: string;
+  address?: 42;
+  connectionId?: number;
+  options?: 42 | cmapPoolOptions;
+  reason: string;
+};
+
+const knownTestKeys = [
+  'name',
+  'version',
+  'style',
+  'description',
+  'poolOptions',
+  'operations',
+  'error',
+  'events',
+  'ignore'
+];
+export type cmapTest = {
+  name?: string; // filename path added by the spec loader
+  version: number;
+  style: 'unit';
+  description: string;
+  poolOptions?: cmapPoolOptions;
+  operations: cmapOperation[];
+  error?: {
+    type: string;
+    message: string;
+    address?: number;
+  };
+  events?: cmapEvent[];
+  ignore?: string[];
+};
+
+const ALL_POOL_EVENTS = new Set([
+  ConnectionPool.CONNECTION_POOL_CREATED,
+  ConnectionPool.CONNECTION_POOL_CLOSED,
+  ConnectionPool.CONNECTION_POOL_CLEARED,
+  ConnectionPool.CONNECTION_CREATED,
+  ConnectionPool.CONNECTION_READY,
+  ConnectionPool.CONNECTION_CLOSED,
+  ConnectionPool.CONNECTION_CHECK_OUT_STARTED,
+  ConnectionPool.CONNECTION_CHECK_OUT_FAILED,
+  ConnectionPool.CONNECTION_CHECKED_OUT,
+  ConnectionPool.CONNECTION_CHECKED_IN
+]);
+
+function getEventType(event) {
+  const eventName = event.constructor.name;
+  return eventName.substring(0, eventName.lastIndexOf('Event'));
+}
+
+/**
+ * In the cmap spec and runner definition,
+ * a "thread" refers to a concurrent execution context
+ */
+class Thread {
+  _promise: Promise<void>;
+  _error: Error;
+  _killed = false;
+
+  _knownCommands: any;
+
+  // concurrent execution context
+  constructor(operations) {
+    this._promise = new Promise(resolve => {
+      this.start = () => resolve();
+    });
+
+    this._knownCommands = operations;
+  }
+
+  start: () => void;
+
+  queue(op: cmapOperation) {
+    if (this._killed || this._error) {
+      return;
+    }
+
+    this._promise = this._promise.then(() => this._runOperation(op)).catch(e => (this._error = e));
+  }
+
+  async _runOperation(op: cmapOperation) {
+    const operationFn = this._knownCommands[op.name];
+    if (!operationFn) {
+      throw new Error(`Invalid command ${op.name}`);
+    }
+
+    await operationFn(op);
+    await sleep();
+  }
+
+  async finish() {
+    this._killed = true;
+    await this._promise;
+    if (this._error) {
+      throw this._error;
+    }
+  }
+}
+
+/**
+ * Implements the spec test match function, see:
+ * [CMAP Spec Test README](https://github.com/mongodb/specifications/tree/master/source/connection-monitoring-and-pooling/tests#spec-test-match-function)
+ */
+const compareInputToSpec = (input, expected) => {
+  if (expected === 42) {
+    expect(input).to.be.ok; // not null or undefined
+    return;
+  }
+
+  if (Array.isArray(expected)) {
```

Comment:
> Assert that actual is of the same JSON type as expected

Not sure this applies to us 🤔 should we be doing something like `expect(JSON.parse(JSON.stringify(actual)).to.equal.(expected)`? I'm thinking of the edge cases where NaN turns into null or something

Reply:
no, the final assertion check already takes care of it; for an array, we do check explicitly, for everything else, it's going to iterate through it down to the last key which has to be a primitive type, where a reference check is made, you can't have two references equal each other if they aren't of the same type; if you have an input that is NaN for the "expected" then that last assertion will always fail; if you have an input that is a NaN, again, it won't pass the check